<a href="https://colab.research.google.com/github/Osrock626/PC_Python_2025II/blob/main/clase_2/conceptos_basicos_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Programación de Computadores en Python
## Introducción a conceptos básicos
#### Universidad Nacional de Colombia
---
**Objetivo:** Aprender los fundamentos iniciales de Python, incluyendo el uso de `print`, creación de variables, tipos de datos y operaciones básicas.

### 📑 Contenido
1. Sentencia `print`
2. Creación de variables
3. Tipos de datos
4. Operaciones entre variables

---
💡 **Tip:** Recuerda que puedes ejecutar cada celda con `Shift + Enter` en Google Colab.

## 1️⃣ Sentencia `print`
La función `print()` sirve para mostrar información en pantalla.

**Sintaxis básica:**
```python
print(valor_o_texto)
```

Puedes imprimir texto, números o el resultado de operaciones.

In [ ]:
# Ejemplos de print
print("Hola Mundo")
print(42)
print("El resultado de 2+3 es:", 2+3)

Hola Mundo
42
El resultado de 2+3 es: 5


### 📝 Ejercicio 1
1. Imprime tu nombre.
2. Imprime tu edad.
3. Imprime la suma de 15 + 8.

In [19]:
# Escribe tu código aquí
print("Wasa mai niga.")
print("Oscar")
print(20)
print(15+8)

Wasa mai niga.
Oscar
20
23


💡 **Curiosidad:** En Python 3, `print` es una función y necesita paréntesis. En versiones antiguas (Python 2) no era así.

## 🖋️ Otras formas de hacer `print` con formateo

En Python, no solo puedes imprimir texto de forma directa, también puedes **formatear cadenas** para mostrar variables y resultados de forma más elegante y legible.  

---

### 1️⃣ Concatenación con `+`
Une cadenas con `+`. Si vas a unir texto con números, debes convertir los números a cadena con `str()`.  
```python
nombre = "Ana"
edad = 25
print("Hola " + nombre + ", tienes " + str(edad) + " años.")
```
### 2️⃣ Comas en `print`
Puedes imprimir varios valores separándolos con comas dentro de `print()`. Python convertirá cada elemento a cadena (si hace falta) y los separará con espacios por defecto.
```python
nombre = "Ana"
edad = 25
print("Hola", nombre, "tienes", edad, "años.")
```
### 3️⃣ Método `.format()`
El método `.format()` permite insertar valores en posiciones `{}` dentro de una cadena de texto.  
Puedes usarlo de forma simple o asignando nombres/índices para mayor claridad.
```python
nombre = "Ana"
edad = 25
print("Hola {}, tienes {} años.".format(nombre, edad))
```
### 4️⃣ F-Strings (recomendado en Python 3.6+)
Las **f-strings** son la forma más cómoda y legible de formatear cadenas en Python.  
Se escriben anteponiendo `f` a la cadena y colocando variables o expresiones dentro de `{}`.
```python
nombre = "Ana"
edad = 25
print(f"Hola {nombre}, tienes {edad} años.")
```

## 2️⃣ Creación de variables
Una variable es un espacio donde puedes guardar un valor para usarlo más adelante.

**Reglas:**
- El nombre debe comenzar con letra o guion bajo.
- No puede contener espacios.
- No debe tener caracteres especiales, salvo el guion bajo _
- No usar palabras reservadas de Python.
- Las variables son sensibles a las mayúsculas, es decir, A es diferente de a

**Sintaxis:**
```python
nombre_variable = valor
```

### palabras reservadas
Python tiene un conjunto de palabras reservadas que no puedes utilizar como nombre de variables:

In [18]:
help("keywords")


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield
break               for                 not                 



In [28]:
# Ejemplo de variables
nombre = "Carlos"
edad = 20
pi = 3.1416

#print(nombre)
#print("Edad:", edad)
#print("Valor de pi:", pi)
#print(nombre, "Edad:", edad, "Valor de PI:", pi)
print(f"Me llamo {nombre},\nTengo: {edad},\nValor de PI:, pi")

print(f"El dijo \"Mi nombre es {nombre}\", Tambien dijo \"Tengo {edad} años.\", y finalmente dijo que: \"Pi tiene el valor {pi}.\"")

print(f"El dijo \"\\\" XD, osea slash en reversa")

print(f"El dijo \"Mi nombre es {nombre},\\\nmi edad es {edad},\\\ny pi es {pi}\"")

Me llamo Carlos,
Tengo: 20,
Valor de PI:, pi
El dijo "Mi nombre es Carlos", Tambien dijo "Tengo 20 años.", y finalmente dijo que: "Pi tiene el valor 3.1416."
El dijo "\" XD, osea slash en reversa
El dijo "Mi nombre es Carlos,\
mi edad es 20,\
y pi es 3.1416"


### 📝 Ejercicio 2
1. Crea una variable con tu color favorito.
2. Crea una variable con tu año de nacimiento.
3. Imprime ambas variables.

In [30]:
# Escribe tu código aquí
color = "Rojo"
año = 2004

print(color,año)
print(color+" "+str(año))
print(color+' '+str(año))

Rojo 2004
Rojo 2004
Rojo 2004


💡 **Tip:** En Python no necesitas declarar el tipo de variable antes de usarla pero es una buena práctica.

## 3️⃣ Tipos de datos
Python tiene varios tipos de datos básicos:
- **int**: números enteros (`5`, `-3`, `100`)
- **float**: números decimales (`3.14`, `-0.5`)
- **complex**: números complejos (`2+5j`, `5+7j`)
- **str**: cadenas de texto (`"Hola"`, `'Python'`)
- **bool**: valores lógicos (`True`, `False`)
- **NoneType**: valores nulos (`None`)

Puedes saber el tipo de una variable con `type()`.

In [34]:
# Ejemplo de tipos de datos
a = 10
b = 3.5
c = 2+5j
d = "Python"
e = True
f = None
g = input("Ingrese numero: ")

print(type(a))
print(type(b))
print(type(c))
print(type(d))
print(type(e))
print(type(f))
print("El dato ingresado es:",type(g))

Ingrese numero: 4
<class 'int'>
<class 'float'>
<class 'complex'>
<class 'str'>
<class 'bool'>
<class 'NoneType'>
El dato ingresado es: <class 'str'>


#### Notación científica para los flotantes

In [36]:
flotante = 0.0005
flotante_not_cient = 5e-4
print(f"El flotante {flotante} en notación científica es {flotante_not_cient}")

El flotante 0.0005 en notación científica es 0.0005


In [37]:
flotante = 2500.0
flotante_not_cient = 25e2
print(f"El flotante {flotante} en notación científica es {flotante_not_cient}")

El flotante 2500.0 en notación científica es 2500.0


## Especificar el tipo de dato al crear variables

En Python, el **tipo de dato** normalmente se asigna automáticamente dependiendo del valor que le des a la variable (esto se llama *tipado dinámico*).  
Sin embargo, también puedes **indicar explícitamente** el tipo de dato que quieres, lo cual es útil para mayor claridad o control.

---

### 📌 Asignación automática (tipado dinámico)
```python
x = 10         # int
y = 3.14       # float
nombre = "Ana" # str
activo = True  # bool
```

### 📌 Conversión explícita (casting)
Puedes usar funciones constructoras para especificar el tipo desde el inicio:
```python
edad = int(25)        # entero
altura = float(1.75)  # decimal
nombre = str("Luis")  # cadena
activo = bool(True)   # booleano
```

### 📌 Anotaciones de tipo (type hints)
Desde Python 3.5+, puedes anotar el tipo esperado.
Esto no fuerza el tipo, pero ayuda a documentar el código y trabajar mejor con herramientas.
```python
edad: int = 25
altura: float = 1.75
nombre: str = "Luis"
activo: bool = True
```

### 📝 Ejercicio 3
1. Crea 4 variables: una entera, una decimal, una cadena y un booleano.
2. Imprime el valor y el tipo de cada una.

In [39]:
# Escribe tu código aquí
Whole = 3
Floaty = 3.56789
String = "papuseñal"
Boolean = False
Bigga = 1234567890

print(type(Whole),type(Floaty),type(String),type(Boolean),type(Bigga))


<class 'int'> <class 'float'> <class 'str'> <class 'bool'> <class 'int'>


💡 **Dato curioso:** Python puede manejar enteros de cualquier tamaño, incluso números más grandes que en otros lenguajes.

## 4️⃣ Operaciones entre variables
Puedes realizar operaciones matemáticas y de concatenación.

### Operaciones matemáticas con números
- Suma: `+`
- Resta: `-`
- Multiplicación: `*`
- División: `/`
- División entera: `//`
- Módulo (resto): `%`
- Potencia: `**`

### Orden de las operaciones
La ejecución de las operaciones se da según el nivel de precedencia. Cuando hay dos o más operaciones del mismo nivel, se realizan de izquierda a derecha. Para la exponenciación es de derecha a izquierda.

| Tipo                     | Operador | Precedencia |
|--------------------------|----------|-------------|
| Paréntesis               | ()       | 1           |
| Exponenciación           | **       | 2           |
| Cambio de signo          | - +      | 3           |
| Multiplicación           | *        | 4           |
| División                 | /        | 4           |
| División entera          | //       | 4           |
| Módulo (o resto)         | %        | 4           |
| Suma                     | +        | 5           |
| Resta                    | -        | 5           |

In [ ]:
# Ejemplo con números
x = 8
y = 3

print("Suma:", x + y)
print("Resta:", x - y)
print("Multiplicación:", x * y)
print("División:", x / y)
print("División entera:", x // y)
print("Módulo:", x % y)
print("Potencia:", x ** y)

Suma: 11
Resta: 5
Multiplicación: 24
División: 2.6666666666666665
División entera: 2
Módulo: 2
Potencia: 512


### Concatenación de cadenas
Se hace usando `+` para unir cadenas o `*` para repetirlas.

In [ ]:
# Ejemplo de concatenación
texto1 = "Hola"
texto2 = "Mundo"

print(texto1 + " " + texto2)
print(texto1 * 3)

Hola Mundo
HolaHolaHola


### 📝 Ejercicio 4
1. Crea dos números y realiza todas las operaciones básicas.
2. Crea dos cadenas y únelas con un espacio.
3. Repite una cadena 5 veces.

In [43]:
# Escribe tu código aquí
Num1 = 7
Num2 = 14

print((str(Num1)+" "+str(Num2))*5)


7 147 147 147 147 14


## 5️⃣ Entrada y salida
#### Saltos de línea y comillas
* Si queremos imprimir varias líneas de texto agregamos ``\n``

In [ ]:
print("Hoy es", 8, "de abril", "\nDel año", 2025)

Hoy es 8 de abril 
Del año 2025


* Si queremos utilizar comillas dobles usamos ``\"``

In [ ]:
print("\"Así se utilizan las comillas dobles\"")

"Así se utilizan las comillas dobles"


#### Manejo de ``print`` con variables
* Con operador ``%``

In [45]:
print("Numero entero: %d \nNumero: flotante %.2f \nCadena: %s"%(12, 3.1415, "Hola"))

print("Numero entero: %-d \nNumero: flotante %-.2f \nCadena: %-s"%(12, 3.1415, "Hola"))

Numero entero: 12 
Numero: flotante 3.14 
Cadena: Hola
Numero entero: 12 
Numero: flotante 3.14 
Cadena: Hola


### Ingreso de datos
Para ingresar datos y asignarlos a una variable, se debe utilizar la sentencia ``input``.

In [48]:
ingreso = input("Ingrese el país ")
print("Eres de", ingreso)

Ingrese el país Joe Mama
Eres de Joe Mama


También se puede con cualquier tipo de dato.

In [50]:
pais = input("Ingrese el país: ")
fecha_1 = int(input("Ingrese día: "))
fecha_2 = input("Ingrese mes: ")
fecha_3 = int(input("Ingrese año: "))
print(f"En {pais} es {fecha_1} de {fecha_2} del {fecha_3}.")

Ingrese el país: Nigga
Ingrese día: 69
Ingrese mes: 69
Ingrese año: 6969
En Nigga es 69 de 69 del 6969.


In [ ]:
print(type(pais), type(fecha_1), type(fecha_2), type(fecha_3))

<class 'str'> <class 'int'> <class 'str'> <class 'int'>


### Comentarios
* Los comentarios permiten explicar el código, lo cual es importante si se trabaja en equipo.
* Se realizan usualmente con un ``#``, aunque con comillas dobles se puede extender en varias líneas.

In [ ]:
# Este es un comentario
# --- El siguiente código realiza la potencia a^n
# Sojnghiofngohif
a = int(input("Ingrese base: ")) # variable b
n = int(input("Ingrese exponente: "))
print(f"El resultado es {a**n}")

Ingrese base: 5
Ingrese exponente: 6
El resultado es 15625


In [ ]:
"""
Este es un comentario extendido en varias lineas

El siguiente código realiza la potencia a^n
higjohijghio

"""
a = int(input("Ingrese base: "))
n = int(input("Ingrese exponente: "))
print(f"El resultado es {a**n}")

Ingrese base: 5
Ingrese exponente: 6
El resultado es 15625


## 6️⃣ Errores y depuración
### Error de sintáxis
Son errores en la estructura del código que son alertados inmediatamente previos a la ejecución.

In [60]:
suma = (1 + 2
print(suma)

SyntaxError: '(' was never closed (ipython-input-1230951632.py, line 1)

### Error de tiempo de ejecución (runtime)
Son errores que solo son alertados cuando el código es ejecutado.

In [61]:
numerador = 10
denominador = 0
resultado = numerador / denominador  # Genera un error ZeroDivisionError en la ejecución
print(resultado)

ZeroDivisionError: division by zero

### Error semántico
Son errores de significado que el programador realiza en el código que no son alertados en la ejecución y pasan desapercibidos.

In [62]:
a = 10
b = 6
resultado = a * b
print(f"El resultado de la suma es: {resultado}")  # Esto mostrará 60 en lugar de 16

El resultado de la suma es: 60


## ✏ Ejercicios

#### a) Evalúa a mano las siguientes expresiones, luego verifícalas con código:

* 6 * 5
* 4 ** 3
* 62 / 13
* 62 // 13
* 12 % 12
* 36 % 5

In [ ]:
"""
Aquí debajo escribe tu solución c:
30
64
4.769
4
0
1
"""


#### b) Evalúa el orden de las siguientes operaciones matemáticas, luego verifícalas con código:

* 4**2 - (9 / 3) + 1
* (3 + 15) / 6 * 2**3
* 20 % 3 + 5 * (40 // 7)

In [51]:
"""
Aquí debajo escribe tu solución c:
12
24
27
"""
print(4**2 - (9 / 3) + 1)
print((3 + 15) / 6 * 2**3)
print(20 % 3 + 5 * (40 // 7))

14.0
24.0
27


#### c) Calcula el volumen de una esfera con un diametro $d = 9 \text{ cm}$ utilizando la fórmula de $V = \frac{4}{3} \pi r^{3}$

In [ ]:
"""
Aquí debajo escribe tu solución c:
"""


#### d) Supón que el precio en una tienda en línea para un libro de cálculo es de \$125.000, pero se encuentra en oferta con un descuento del 25\%. Pero la tienda hace un cargo para el envío de \$5.000 para la primera unidad y de \$500 para cada adicional. ¿Cuál es el costo total para 35 copias?

In [58]:
"""
Aquí debajo escribe tu solución c:
"""
n = int(input())
p = 125000
costo = (n-1)*500 + n*p*.75 + 5000
print(costo)



35
3303250.0
